File: src/main/kotlin/demo/language/OverloadsKotlin.java
package demo.language;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.stream.Collectors;
import java.util.stream.Stream;

// TODO: Convert to Kotlin
public class OverloadsKotlin {

    public void query(@NotNull String from) {
        query("*", from);
    }

    public void query(@NotNull String select, @NotNull String from) {
        query(select, from, null);
    }

    public void query(@NotNull String select, @NotNull String from, @Nullable String where
    ) {
        query(select, from, where, null);
    }

    public void query(@NotNull String select, @NotNull String from, @Nullable String where, @Nullable String groupBy) {
        query(select, from, where, groupBy, null);
    }

    public void query(
        @NotNull String select,
        @NotNull String from,
        @Nullable String where,
        @Nullable String groupBy,
        @Nullable String orderBy
    ) {
        String query = Stream.of(select, from, where, groupBy, orderBy)
            .map(it -> it != null ? it : "_")
            .collect(Collectors.joining(";"));
        System.out.println(query);
    }


    public static void main(String[] args) {
        OverloadsKotlin overloads = new OverloadsKotlin();
        overloads.query("users");
        overloads.query("name", "users");
        overloads.query("name", "users", null, null, "name asc");
    }
}

File: src/main/kotlin/demo/language/OverloadsKotlin.kt
package demo.language

import java.util.stream.Collectors
import java.util.stream.Stream

// TODO: Convert to Kotlin
class OverloadsKotlin {

    fun query(
        select: String = "*",
        from: String,
        where: String? = null,
        groupBy: String? = null,
        orderBy: String? = null
    ) {
        val query = Stream.of<String>(select, from, where, groupBy, orderBy)
            .map { it -> it ?: "_" }
            .collect(Collectors.joining(";"))
        println(query)
    }

    companion object {


        @JvmStatic
        fun main(args: Array<String>) {
            val overloads = OverloadsKotlin()
            overloads.query(from = "users")
            overloads.query("name", "users")
            overloads.query(select = "name", from = "users", orderBy = "name asc")
        }
    }
}
